bliss/unconstrained_dists.py

```diff
@@ -40,26 +40,43 @@ def __init__(self, mu, sigma):
         super().__init__(validate_args=False)
 
         multiple_normals = Normal(mu, sigma)
-        prob_in_unit_box_hw = multiple_normals.cdf(torch.ones_like(mu))
-        prob_in_unit_box_hw -= multiple_normals.cdf(torch.zeros_like(mu))
-        self.log_prob_in_unit_box = prob_in_unit_box_hw.log().sum(dim=-1)
-
         self.base_dist = Independent(multiple_normals, 1)
 
+        prob_in_unit_box_hw = multiple_normals.cdf(self.b) - multiple_normals.cdf(self.a)
+        self.log_prob_in_unit_box = prob_in_unit_box_hw.log().sum(dim=-1)
+
     def __repr__(self):
         return f"{self.__class__.__name__}({self.base_dist.base_dist})"
 
-    def sample(self, **args):
-        # some ideas for how to sample it here, if we need to:
-        # https://cran.r-project.org/web/packages/truncnorm/
-        raise NotImplementedError("sampling a truncated normal isn't straightforward")
+    def sample(self, sample_shape=(1,), **args):
+        """Generate sample using rejection sampling."""
+        q = Independent(Normal(self.base_dist.mean, self.base_dist.stddev), 1)
+        samples = q.sample(sample_shape)
+        valid = (samples.min(dim=-1)[0] >= 0) & (samples.max(dim=-1)[0] < 1)
+        while not valid.all():
+            new_samples = q.sample(sample_shape)
+            samples[~valid] = new_samples[~valid]
+            valid = (samples.min(dim=-1)[0] >= 0) & (samples.max(dim=-1)[0] < 1)
+        return samples
+
+    @property
+    def a(self):
+        return torch.zeros_like(self.base_dist.mean)
+
+    @property
+    def b(self):
+        return torch.ones_like(self.base_dist.mean)
 
     @property
     def mean(self):
-        return self.base_dist.mean
+        mu = self.base_dist.mean
+        offset = self.base_dist.log_prob(self.a).exp() - self.base_dist.log_prob(self.b).exp()
+        offset /= self.log_prob_in_unit_box.exp()
+        return mu + offset.unsqueeze(-1)
 
     @property
     def stddev(self):
+        # TODO: add offset term https://en.wikipedia.org/wiki/Truncated_normal_distribution
```

Comment:
How about raising a `NotImplemented` error if this function is called?

Reply:
Sounds good, I had left it in case we need a rough estimate but we can always just get that from `base_dist.stddev` I guess.

Reply:
Yeah I'm wary of returning a value that's close but incorrect...could be difficult to detect a problem